homeassistant/components/humidifier/__init__.py

```diff
@@ -0,0 +1,308 @@
+    OPERATION_MODES,
+    SERVICE_SET_FAN_MODE,
+    SERVICE_SET_HUMIDITY,
+    SERVICE_SET_OPERATION_MODE,
+    SERVICE_SET_PRESET_MODE,
+    SUPPORT_FAN_MODE,
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_HUMIDITY,
+    SUPPORT_TEMPERATURE,
+    SUPPORT_WATER_LEVEL,
+    DEFAULT_MIN_HUMIDITY,
+    DEFAULT_MAX_HUMIDITY,
+)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+SCAN_INTERVAL = timedelta(seconds=60)
+
+_LOGGER = logging.getLogger(__name__)
+
+SET_FAN_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_FAN_MODE): cv.string}
+)
+SET_PRESET_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_PRESET_MODE): cv.string}
+)
+SET_OPERATION_MODE_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_OPERATION_MODE): vol.In(OPERATION_MODES)}
+)
+SET_HUMIDITY_SCHEMA = ENTITY_SERVICE_SCHEMA.extend(
+    {vol.Required(ATTR_HUMIDITY): vol.Coerce(float)}
+)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up humidifier devices."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_TURN_ON, ENTITY_SERVICE_SCHEMA, "async_turn_on"
+    )
+    component.async_register_entity_service(
+        SERVICE_TURN_OFF, ENTITY_SERVICE_SCHEMA, "async_turn_off"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_FAN_MODE, SET_FAN_MODE_SCHEMA, "async_set_fan_mode"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_OPERATION_MODE,
+        SET_OPERATION_MODE_SCHEMA,
+        "async_set_operation_mode",
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_PRESET_MODE, SET_PRESET_MODE_SCHEMA, "async_set_preset_mode"
+    )
+    component.async_register_entity_service(
+        SERVICE_SET_HUMIDITY, SET_HUMIDITY_SCHEMA, "async_set_humidity"
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry):
+    """Set up a config entry."""
+    return await hass.data[DOMAIN].async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry):
+    """Unload a config entry."""
+    return await hass.data[DOMAIN].async_unload_entry(entry)
+
+
+class HumidifierDevice(Entity):
+    """Representation of a humidifier device."""
+
+    @property
+    def state(self) -> str:
+        """Return the current state."""
+        return self.operation_mode
+
+    @property
+    def state_attributes(self) -> Dict[str, Any]:
+        """Return the optional state attributes."""
+        supported_features = self.supported_features
+        data = {
+            ATTR_OPERATION_MODES: self.operation_modes,
+            ATTR_CURRENT_HUMIDITY: self.current_humidity,
+        }
+
+        if supported_features & SUPPORT_TARGET_HUMIDITY:
+            data[ATTR_HUMIDITY] = self.target_humidity
+            data[ATTR_MIN_HUMIDITY] = self.min_humidity
```

Comment:
These should be moved to `capability_attributes` (which is in latest dev)

Reply:
done